src/day10.scala

```diff
@@ -0,0 +1,118 @@
+// using scala 3.1.0
+
+package day10
+
+import scala.util.Using
+import scala.io.Source
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day10"))(_.mkString)
+
+enum CheckResult:
+  case Ok
+  case IllegalClosing(expected: Option[Symbol], found: Symbol)
+  case Incomplete(pending: List[Symbol])
+
+extension (illegalClosing: CheckResult.IllegalClosing)
+  def score: Int = 
+    import Symbol.*
+    illegalClosing.found match
+      case Parenthesis(_) => 3
+      case Bracket(_) => 57
+      case Brace(_) => 1197
+      case Diamond(_) => 25137
+
+enum SymbolState:
+  case Open, Close
+
+enum Symbol(val state: SymbolState):
+  case Parenthesis(override val state: SymbolState) extends Symbol(state)
+  case Bracket(override val state: SymbolState) extends Symbol(state)
+  case Brace(override val state: SymbolState) extends Symbol(state)
+  case Diamond(override val state: SymbolState) extends Symbol(state)
```

Comment:
This syntax looks complicated to me.

I wonder if the following works:

```suggestion
enum Symbol:
  def state: SymbolState
  case Parenthesis(state: SymbolState)
  case Bracket(state: SymbolState)
  case Brace(state: SymbolState)
  case Diamond(state: SymbolState)
```

If it doesn’t, just fallback to a trait:

```suggestion
sealed trait Symbol:
  def state: SymbolState
case Parenthesis(state: SymbolState) extends Symbol
case Bracket(state: SymbolState) extends Symbol
case Brace(state: SymbolState) extends Symbol
case Diamond(state: SymbolState) extends Symbol
```

Both suggestion look lighter to my eyes.